In [1]:
from scipy.io import arff
import pandas as pd

In [2]:
data = arff.loadarff(r"E:\ZIP1--clustering-benchmark-master\clustering-benchmark-master\src\main\resources\datasets\real-world\dermatology.arff")

In [3]:
df = pd.DataFrame(data[0])
df.head()

erythema scaling definite_borders itching koebner_phenomenon  \
0     b'2'    b'2'             b'0'    b'3'               b'0'   
1     b'3'    b'3'             b'3'    b'2'               b'1'   
2     b'2'    b'1'             b'2'    b'3'               b'1'   
3     b'2'    b'2'             b'2'    b'0'               b'0'   
4     b'2'    b'3'             b'2'    b'2'               b'2'   

  polygonal_papules follicular_papules oral_mucosal_involvement  \
0              b'0'               b'0'                     b'0'   
1              b'0'               b'0'                     b'0'   
2              b'3'               b'0'                     b'3'   
3              b'0'               b'0'                     b'0'   
4              b'2'               b'0'                     b'2'   

  knee_and_elbow_involvement scalp_involvement  ...  \
0                       b'1'              b'0'  ...   
1                       b'1'              b'1'  ...   
2                       b'0'              b'0'  ...   
3                       b'3'              b'2'  ...   
4                       b'0'              b'0'  ...   

  disappearance_of_the_granular_layer vacuolisation_and_damage_of_basal_layer  \
0                                b'0'                                    b'0'   
1                                b'0'                                    b'0'   
2                                b'0'                                    b'2'   
3                                b'3'                                    b'0'   
4                                b'2'                                    b'3'   

  spongiosis saw-tooth_appearance_of_retes follicular_horn_plug  \
0       b'3'                          b'0'                 b'0'   
1       b'0'                          b'0'                 b'0'   
2       b'3'                          b'2'                 b'0'   
3       b'0'                          b'0'                 b'0'   
4       b'2'                          b'3'                 b'0'   

  perifollicular_parakeratosis inflammatory_monoluclear_inflitrate  \
0                         b'0'                                b'1'   
1                         b'0'                                b'1'   
2                         b'0'                                b'2'   
3                         b'0'                                b'3'   
4                         b'0'                                b'2'   

  band-like_infiltrate   Age class  
0                 b'0'  55.0  b'2'  
1                 b'0'   8.0  b'1'  
2                 b'3'  26.0  b'3'  
3                 b'0'  40.0  b'1'  
4                 b'3'  45.0  b'3'  

[5 rows x 35 columns]

In [4]:
df.to_csv('dermatology.csv', index = False)

In [5]:
df1 = pd.read_csv('dermatology.csv')
df1

erythema scaling definite_borders itching koebner_phenomenon  \
0       b'2'    b'2'             b'0'    b'3'               b'0'   
1       b'3'    b'3'             b'3'    b'2'               b'1'   
2       b'2'    b'1'             b'2'    b'3'               b'1'   
3       b'2'    b'2'             b'2'    b'0'               b'0'   
4       b'2'    b'3'             b'2'    b'2'               b'2'   
..       ...     ...              ...     ...                ...   
361     b'2'    b'1'             b'1'    b'0'               b'1'   
362     b'3'    b'2'             b'1'    b'0'               b'1'   
363     b'3'    b'2'             b'2'    b'2'               b'3'   
364     b'2'    b'1'             b'3'    b'1'               b'2'   
365     b'3'    b'2'             b'2'    b'0'               b'0'   

    polygonal_papules follicular_papules oral_mucosal_involvement  \
0                b'0'               b'0'                     b'0'   
1                b'0'               b'0'                     b'0'   
2                b'3'               b'0'                     b'3'   
3                b'0'               b'0'                     b'0'   
4                b'2'               b'0'                     b'2'   
..                ...                ...                      ...   
361              b'0'               b'0'                     b'0'   
362              b'0'               b'0'                     b'0'   
363              b'2'               b'0'                     b'2'   
364              b'3'               b'0'                     b'2'   
365              b'0'               b'0'                     b'0'   

    knee_and_elbow_involvement scalp_involvement  ...  \
0                         b'1'              b'0'  ...   
1                         b'1'              b'1'  ...   
2                         b'0'              b'0'  ...   
3                         b'3'              b'2'  ...   
4                         b'0'              b'0'  ...   
..                         ...               ...  ...   
361                       b'0'              b'0'  ...   
362                       b'0'              b'0'  ...   
363                       b'0'              b'0'  ...   
364                       b'0'              b'0'  ...   
365                       b'3'              b'3'  ...   

    disappearance_of_the_granular_layer  \
0                                  b'0'   
1                                  b'0'   
2                                  b'0'   
3                                  b'3'   
4                                  b'2'   
..                                  ...   
361                                b'0'   
362                                b'1'   
363                                b'0'   
364                                b'0'   
365                                b'2'   

    vacuolisation_and_damage_of_basal_layer spongiosis  \
0                                      b'0'       b'3'   
1                                      b'0'       b'0'   
2                                      b'2'       b'3'   
3                                      b'0'       b'0'   
4                                      b'3'       b'2'   
..                                      ...        ...   
361                                    b'0'       b'1'   
362                                    b'0'       b'1'   
363                                    b'3'       b'0'   
364                                    b'2'       b'0'   
365                                    b'0'       b'0'   

    saw-tooth_appearance_of_retes follicular_horn_plug  \
0                            b'0'                 b'0'   
1                            b'0'                 b'0'   
2                            b'2'                 b'0'   
3                            b'0'                 b'0'   
4                            b'3'                 b'0'   
..                            ...                  ...   
361                          b'0'                 b'0

In [6]:
df1.columns


Index(['erythema', 'scaling', 'definite_borders', 'itching',
       'koebner_phenomenon', 'polygonal_papules', 'follicular_papules',
       'oral_mucosal_involvement', 'knee_and_elbow_involvement',
       'scalp_involvement', 'family_history', 'melanin_incontinence',
       'eosinophils_in_the_infiltrate', 'PNL_infiltrate',
       'fibrosis_of_the_papillary_dermis', 'exocytosis', 'acanthosis',
       'hyperkeratosis', 'parakeratosis', 'clubbing_of_the_rete_ridges',
       'elongation_of_the_rete_ridges',
       'thinning_of_the_suprapapillary_epidermis', 'spongiform_pustule',
       'munro_microabcess', 'focal_hypergranulosis',
       'disappearance_of_the_granular_layer',
       'vacuolisation_and_damage_of_basal_layer', 'spongiosis',
       'saw-tooth_appearance_of_retes', 'follicular_horn_plug',
       'perifollicular_parakeratosis', 'inflammatory_monoluclear_inflitrate',
       'band-like_infiltrate', 'Age', 'class'],
      dtype='object')

In [7]:
lt=['erythema', 'scaling','definite_borders', 'itching',
       'koebner_phenomenon', 'polygonal_papules', 'follicular_papules',
       'oral_mucosal_involvement', 'knee_and_elbow_involvement',
       'scalp_involvement', 'family_history', 'melanin_incontinence',
       'eosinophils_in_the_infiltrate', 'PNL_infiltrate',
       'fibrosis_of_the_papillary_dermis', 'exocytosis', 'acanthosis',
       'hyperkeratosis', 'parakeratosis', 'clubbing_of_the_rete_ridges',
       'elongation_of_the_rete_ridges',
       'thinning_of_the_suprapapillary_epidermis', 'spongiform_pustule',
       'munro_microabcess', 'focal_hypergranulosis',
       'disappearance_of_the_granular_layer',
       'vacuolisation_and_damage_of_basal_layer', 'spongiosis',
       'saw-tooth_appearance_of_retes', 'follicular_horn_plug',
       'perifollicular_parakeratosis', 'inflammatory_monoluclear_inflitrate',
       'band-like_infiltrate',]

In [8]:
df1['corrected_erythema'] = df1['erythema'].apply(lambda x : x.split("'")[1])
df1['corrected_scaling'] = df1['scaling'].apply(lambda x : x.split("'")[1])
df1['corrected_definite_borders'] = df1['definite_borders'].apply(lambda x : x.split("'")[1])
df1['corrected_itching'] = df1['itching'].apply(lambda x : x.split("'")[1])
df1['corrected_koebner_phenomenon'] = df1['koebner_phenomenon'].apply(lambda x : x.split("'")[1])
df1['corrected_polygonal_papules'] = df1['polygonal_papules'].apply(lambda x : x.split("'")[1])
df1['corrected_follicular_papules'] = df1['follicular_papules'].apply(lambda x : x.split("'")[1])
df1['corrected_oral_mucosal_involvement'] = df1['oral_mucosal_involvement'].apply(lambda x : x.split("'")[1])
df1['corrected_knee_and_elbow_involvement'] = df1['knee_and_elbow_involvement'].apply(lambda x : x.split("'")[1])
df1['corrected_scalp_involvement'] = df1['scalp_involvement'].apply(lambda x : x.split("'")[1])
df1['corrected_family_history'] = df1['family_history'].apply(lambda x : x.split("'")[1])
df1['corrected_melanin_incontinence'] = df1['melanin_incontinence'].apply(lambda x : x.split("'")[1])
df1['corrected_eosinophils_in_the_infiltrate'] = df1['eosinophils_in_the_infiltrate'].apply(lambda x : x.split("'")[1])
df1['corrected_PNL_infiltrate'] = df1['PNL_infiltrate'].apply(lambda x : x.split("'")[1])
df1['corrected_fibrosis_of_the_papillary_dermis'] = df1['fibrosis_of_the_papillary_dermis'].apply(lambda x : x.split("'")[1])
df1['corrected_exocytosis'] = df1['exocytosis'].apply(lambda x : x.split("'")[1])
df1['corrected_acanthosis'] = df1['acanthosis'].apply(lambda x : x.split("'")[1])
df1['corrected_hyperkeratosis'] = df1['hyperkeratosis'].apply(lambda x : x.split("'")[1])
df1['corrected_parakeratosis'] = df1['parakeratosis'].apply(lambda x : x.split("'")[1])
df1['corrected_clubbing_of_the_rete_ridges'] = df1['clubbing_of_the_rete_ridges'].apply(lambda x : x.split("'")[1])
df1['corrected_elongation_of_the_rete_ridges'] = df1['elongation_of_the_rete_ridges'].apply(lambda x : x.split("'")[1])
df1['corrected_thinning_of_the_suprapapillary_epidermis'] = df1['thinning_of_the_suprapapillary_epidermis'].apply(lambda x : x.split("'")[1])
df1['corrected_spongiform_pustule'] = df1['spongiform_pustule'].apply(lambda x : x.split("'")[1])
df1['corrected_munro_microabcess'] = df1['munro_microabcess'].apply(lambda x : x.split("'")[1])
df1['corrected_focal_hypergranulosis'] = df1['focal_hypergranulosis'].apply(lambda x : x.split("'")[1])
df1['corrected_disappearance_of_the_granular_layer'] = df1['disappearance_of_the_granular_layer'].apply(lambda x : x.split("'")[1])
df1['corrected_vacuolisation_and_damage_of_basal_layer'] = df1['vacuolisation_and_damage_of_basal_layer'].apply(lambda x : x.split("'")[1])
df1['corrected_spongiosis'] = df1['spongiosis'].apply(lambda x : x.split("'")[1])
df1['corrected_saw-tooth_appearance_of_retes'] = df1['saw-tooth_appearance_of_retes'].apply(lambda x : x.split("'")[1])
df1['corrected_follicular_horn_plug'] = df1['follicular_horn_plug'].apply(lambda x : x.split("'")[1])
df1['corrected_perifollicular_parakeratosis'] = df1['perifollicular_parakeratosis'].apply(lambda x : x.split("'")[1])
df1['corrected_inflammatory_monoluclear_inflitrate'] = df1['inflammatory_monoluclear_inflitrate'].apply(lambda x : x.split("'")[1])
df1['corrected_band-like_infiltrate'] = df1['band-like_infiltrate'].apply(lambda x : x.split("'")[1])
df1['corrected_class'] = df1['class'].apply(lambda x : x.split("'")[1])

df1

erythema scaling definite_borders itching koebner_phenomenon  \
0       b'2'    b'2'             b'0'    b'3'               b'0'   
1       b'3'    b'3'             b'3'    b'2'               b'1'   
2       b'2'    b'1'             b'2'    b'3'               b'1'   
3       b'2'    b'2'             b'2'    b'0'               b'0'   
4       b'2'    b'3'             b'2'    b'2'               b'2'   
..       ...     ...              ...     ...                ...   
361     b'2'    b'1'             b'1'    b'0'               b'1'   
362     b'3'    b'2'             b'1'    b'0'               b'1'   
363     b'3'    b'2'             b'2'    b'2'               b'3'   
364     b'2'    b'1'             b'3'    b'1'               b'2'   
365     b'3'    b'2'             b'2'    b'0'               b'0'   

    polygonal_papules follicular_papules oral_mucosal_involvement  \
0                b'0'               b'0'                     b'0'   
1                b'0'               b'0'                     b'0'   
2                b'3'               b'0'                     b'3'   
3                b'0'               b'0'                     b'0'   
4                b'2'               b'0'                     b'2'   
..                ...                ...                      ...   
361              b'0'               b'0'                     b'0'   
362              b'0'               b'0'                     b'0'   
363              b'2'               b'0'                     b'2'   
364              b'3'               b'0'                     b'2'   
365              b'0'               b'0'                     b'0'   

    knee_and_elbow_involvement scalp_involvement  ...  \
0                         b'1'              b'0'  ...   
1                         b'1'              b'1'  ...   
2                         b'0'              b'0'  ...   
3                         b'3'              b'2'  ...   
4                         b'0'              b'0'  ...   
..                         ...               ...  ...   
361                       b'0'              b'0'  ...   
362                       b'0'              b'0'  ...   
363                       b'0'              b'0'  ...   
364                       b'0'              b'0'  ...   
365                       b'3'              b'3'  ...   

    corrected_focal_hypergranulosis  \
0                                 0   
1                                 0   
2                                 2   
3                                 0   
4                                 2   
..                              ...   
361                               0   
362                               0   
363                               3   
364                               3   
365                               0   

    corrected_disappearance_of_the_granular_layer  \
0                                               0   
1                                               0   
2                                               0   
3                                               3   
4                                               2   
..                                            ...   
361                                             0   
362                                             1   
363                                             0   
364                                             0   
365                                             2   

    corrected_vacuolisation_and_damage_of_basal_layer corrected_spongiosis  \
0                                                   0                    3   
1                                                   0                    0   
2                                                   2                    3   
3                                                   0                    0   
4                                                   3                    2   
..                                                ...                  ...   
361                  

In [9]:
df1.columns


Index(['erythema', 'scaling', 'definite_borders', 'itching',
       'koebner_phenomenon', 'polygonal_papules', 'follicular_papules',
       'oral_mucosal_involvement', 'knee_and_elbow_involvement',
       'scalp_involvement', 'family_history', 'melanin_incontinence',
       'eosinophils_in_the_infiltrate', 'PNL_infiltrate',
       'fibrosis_of_the_papillary_dermis', 'exocytosis', 'acanthosis',
       'hyperkeratosis', 'parakeratosis', 'clubbing_of_the_rete_ridges',
       'elongation_of_the_rete_ridges',
       'thinning_of_the_suprapapillary_epidermis', 'spongiform_pustule',
       'munro_microabcess', 'focal_hypergranulosis',
       'disappearance_of_the_granular_layer',
       'vacuolisation_and_damage_of_basal_layer', 'spongiosis',
       'saw-tooth_appearance_of_retes', 'follicular_horn_plug',
       'perifollicular_parakeratosis', 'inflammatory_monoluclear_inflitrate',
       'band-like_infiltrate', 'Age', 'class', 'corrected_erythema',
       'corrected_scaling', 'corrected_defi

In [10]:
df2 = df1.drop(lt, axis = 'columns')
df2

Age class corrected_erythema corrected_scaling  \
0    55.0  b'2'                  2                 2   
1     8.0  b'1'                  3                 3   
2    26.0  b'3'                  2                 1   
3    40.0  b'1'                  2                 2   
4    45.0  b'3'                  2                 3   
..    ...   ...                ...               ...   
361  25.0  b'4'                  2                 1   
362  36.0  b'4'                  3                 2   
363  28.0  b'3'                  3                 2   
364  50.0  b'3'                  2                 1   
365  35.0  b'1'                  3                 2   

    corrected_definite_borders corrected_itching corrected_koebner_phenomenon  \
0                            0                 3                            0   
1                            3                 2                            1   
2                            2                 3                            1   
3                            2                 0                            0   
4                            2                 2                            2   
..                         ...               ...                          ...   
361                          1                 0                            1   
362                          1                 0                            1   
363                          2                 2                            3   
364                          3                 1                            2   
365                          2                 0                            0   

    corrected_polygonal_papules corrected_follicular_papules  \
0                             0                            0   
1                             0                            0   
2                             3                            0   
3                             0                            0   
4                             2                            0   
..                          ...                          ...   
361                           0                            0   
362                           0                            0   
363                           2                            0   
364                           3                            0   
365                           0                            0   

    corrected_oral_mucosal_involvement  ... corrected_focal_hypergranulosis  \
0                                    0  ...                               0   
1                                    0  ...                               0   
2                                    3  ...                               2   
3                                    0  ...                               0   
4                                    2  ...                               2   
..                                 ...  ...                             ...   
361                                  0  ...                               0   
362                                  0  ...                               0   
363                                  2  ...                               3   
364                                  2  ...                               3   
365                                  0  ...                               0   

    corrected_disappearance_of_the_granular_layer  \
0                                               0   
1                                               0   
2                                               0   
3                                               3   
4                                               2   
..                                            ...   
361                                             0   
362                                             1   
363                                             0   
364                                             0   
365                                             2   

    c

In [11]:
derma = df2.drop(['class'], axis ='columns')
derma.to_csv('derma.csv', index = False)

In [12]:
input_set = df2.drop(['class','corrected_class'], axis ='columns')
input_set

Age corrected_erythema corrected_scaling corrected_definite_borders  \
0    55.0                  2                 2                          0   
1     8.0                  3                 3                          3   
2    26.0                  2                 1                          2   
3    40.0                  2                 2                          2   
4    45.0                  2                 3                          2   
..    ...                ...               ...                        ...   
361  25.0                  2                 1                          1   
362  36.0                  3                 2                          1   
363  28.0                  3                 2                          2   
364  50.0                  2                 1                          3   
365  35.0                  3                 2                          2   

    corrected_itching corrected_koebner_phenomenon  \
0                   3                            0   
1                   2                            1   
2                   3                            1   
3                   0                            0   
4                   2                            2   
..                ...                          ...   
361                 0                            1   
362                 0                            1   
363                 2                            3   
364                 1                            2   
365                 0                            0   

    corrected_polygonal_papules corrected_follicular_papules  \
0                             0                            0   
1                             0                            0   
2                             3                            0   
3                             0                            0   
4                             2                            0   
..                          ...                          ...   
361                           0                            0   
362                           0                            0   
363                           2                            0   
364                           3                            0   
365                           0                            0   

    corrected_oral_mucosal_involvement corrected_knee_and_elbow_involvement  \
0                                    0                                    1   
1                                    0                                    1   
2                                    3                                    0   
3                                    0                                    3   
4                                    2                                    0   
..                                 ...                                  ...   
361                                  0                                    0   
362                                  0                                    0   
363                                  2                                    0   
364                                  2                                    0   
365                                  0                                    3   

     ... corrected_munro_microabcess corrected_focal_hypergranulosis  \
0    ...                           0                               0   
1    ...                           1                               0   
2    ...                           0                               2   
3    ...                           0                               0   
4    ...                           0                               2   
..   ...                         ...                             ...   
361  ...                           0                               0   
362  ...                           0                               0   
363  ...                           0                              

In [13]:
input_set.isna().any()

Age                                                    True
corrected_erythema                                    False
corrected_scaling                                     False
corrected_definite_borders                            False
corrected_itching                                     False
corrected_koebner_phenomenon                          False
corrected_polygonal_papules                           False
corrected_follicular_papules                          False
corrected_oral_mucosal_involvement                    False
corrected_knee_and_elbow_involvement                  False
corrected_scalp_involvement                           False
corrected_family_history                              False
corrected_melanin_incontinence                        False
corrected_eosinophils_in_the_infiltrate               False
corrected_PNL_infiltrate                              False
corrected_fibrosis_of_the_papillary_dermis            False
corrected_exocytosis                    

In [14]:
input_set['Age'] = input_set['Age'].fillna(input_set['Age'].mean())
input_set

Age corrected_erythema corrected_scaling corrected_definite_borders  \
0    55.0                  2                 2                          0   
1     8.0                  3                 3                          3   
2    26.0                  2                 1                          2   
3    40.0                  2                 2                          2   
4    45.0                  2                 3                          2   
..    ...                ...               ...                        ...   
361  25.0                  2                 1                          1   
362  36.0                  3                 2                          1   
363  28.0                  3                 2                          2   
364  50.0                  2                 1                          3   
365  35.0                  3                 2                          2   

    corrected_itching corrected_koebner_phenomenon  \
0                   3                            0   
1                   2                            1   
2                   3                            1   
3                   0                            0   
4                   2                            2   
..                ...                          ...   
361                 0                            1   
362                 0                            1   
363                 2                            3   
364                 1                            2   
365                 0                            0   

    corrected_polygonal_papules corrected_follicular_papules  \
0                             0                            0   
1                             0                            0   
2                             3                            0   
3                             0                            0   
4                             2                            0   
..                          ...                          ...   
361                           0                            0   
362                           0                            0   
363                           2                            0   
364                           3                            0   
365                           0                            0   

    corrected_oral_mucosal_involvement corrected_knee_and_elbow_involvement  \
0                                    0                                    1   
1                                    0                                    1   
2                                    3                                    0   
3                                    0                                    3   
4                                    2                                    0   
..                                 ...                                  ...   
361                                  0                                    0   
362                                  0                                    0   
363                                  2                                    0   
364                                  2                                    0   
365                                  0                                    3   

     ... corrected_munro_microabcess corrected_focal_hypergranulosis  \
0    ...                           0                               0   
1    ...                           1                               0   
2    ...                           0                               2   
3    ...                           0                               0   
4    ...                           0                               2   
..   ...                         ...                             ...   
361  ...                           0                               0   
362  ...                           0                               0   
363  ...                           0                              

In [15]:
input_set.isna().any()

Age                                                   False
corrected_erythema                                    False
corrected_scaling                                     False
corrected_definite_borders                            False
corrected_itching                                     False
corrected_koebner_phenomenon                          False
corrected_polygonal_papules                           False
corrected_follicular_papules                          False
corrected_oral_mucosal_involvement                    False
corrected_knee_and_elbow_involvement                  False
corrected_scalp_involvement                           False
corrected_family_history                              False
corrected_melanin_incontinence                        False
corrected_eosinophils_in_the_infiltrate               False
corrected_PNL_infiltrate                              False
corrected_fibrosis_of_the_papillary_dermis            False
corrected_exocytosis                    

In [16]:
target_set = df1['corrected_class']
target_set

0      2
1      1
2      3
3      1
4      3
      ..
361    4
362    4
363    3
364    3
365    1
Name: corrected_class, Length: 366, dtype: object

In [17]:
from sklearn.model_selection import train_test_split
X,x,Y,y = train_test_split(input_set, target_set, test_size=0.25)


In [18]:
X                 # Age column is now the first column of the dataframe.

Age corrected_erythema corrected_scaling corrected_definite_borders  \
292  27.0                  2                 3                          2   
26   10.0                  1                 1                          0   
209  27.0                  2                 3                          2   
246  55.0                  2                 2                          2   
216  50.0                  2                 2                          3   
..    ...                ...               ...                        ...   
290  27.0                  2                 2                          3   
124  29.0                  3                 2                          2   
191  38.0                  2                 3                          3   
72   52.0                  1                 1                          1   
339  51.0                  2                 1                          2   

    corrected_itching corrected_koebner_phenomenon  \
292                 0                            1   
26                  1                            0   
209                 0                            0   
246                 0                            0   
216                 2                            2   
..                ...                          ...   
290                 2                            1   
124                 2                            2   
191                 3                            0   
72                  1                            1   
339                 2                            2   

    corrected_polygonal_papules corrected_follicular_papules  \
292                           0                            0   
26                            0                            3   
209                           0                            0   
246                           0                            0   
216                           2                            0   
..                          ...                          ...   
290                           2                            0   
124                           0                            0   
191                           3                            0   
72                            0                            0   
339                           3                            0   

    corrected_oral_mucosal_involvement corrected_knee_and_elbow_involvement  \
292                                  0                                    3   
26                                   0                                    1   
209                                  0                                    3   
246                                  0                                    2   
216                                  3                                    0   
..                                 ...                                  ...   
290                                  2                                    0   
124                                  0                                    0   
191                                  3                                    0   
72                                   1                                    0   
339                                  2                                    0   

     ... corrected_munro_microabcess corrected_focal_hypergranulosis  \
292  ...                           0                               0   
26   ...                           0                               1   
209  ...                           1                               0   
246  ...                           2                               0   
216  ...                           0                               2   
..   ...                         ...                             ...   
290  ...                           0                               2   
124  ...                           1                               0   
191  ...                           0                              

In [19]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression( max_iter=1500, random_state=10)
model.fit(X,Y)

LogisticRegression(max_iter=1500, random_state=10)

In [20]:
model.score(x,y)

0.9565217391304348

# Using Cross-validation 

In [21]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

In [22]:
from sklearn.model_selection import cross_val_score
import numpy as np

In [23]:
lr = cross_val_score(LogisticRegression(max_iter=1100), input_set,target_set)
lr

array([0.98648649, 0.97260274, 0.95890411, 0.98630137, 0.93150685])

In [24]:
dt = cross_val_score(DecisionTreeClassifier(), input_set,target_set )
dt

array([0.89189189, 0.95890411, 0.87671233, 0.93150685, 0.95890411])

In [25]:
sv = cross_val_score(SVC(), input_set,target_set )
sv

array([0.66216216, 0.69863014, 0.75342466, 0.7260274 , 0.68493151])

In [26]:
rf = cross_val_score(RandomForestClassifier(n_estimators=40), input_set,target_set)
rf

array([0.98648649, 1.        , 0.97260274, 0.98630137, 0.91780822])

In [27]:
np.average(lr)

0.9671603109959275

In [28]:
np.average(dt)

0.9235838578304332

In [29]:
np.average(sv)

0.7050351721584599

In [30]:
np.average(rf)

0.9726397630507219

# Best score is from Random Forest Classifier which is 97.26% so far.


 Using GridSearchCV without Pipeline

In [31]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import ShuffleSplit

In [47]:
def best_model_using_gridsearchCV(X,Y):
    algos={
        'linear regression' : {
            'model': LogisticRegression(max_iter=2500),
            'parameters' : {
                'fit_intercept' : [True, False],
                'random_state': [1,10,30,50,100],
                'solver': ['liblinear', 'saga','sag']
            }
        },
#         'decision tree': {
#             'model' : DecisionTreeClassifier(),
#             'parameters' :{
#                 'random_state': [1,5,10,30,50,100],
#                 'criterion': ['gini', 'entropy'],
#                 'splitter':['best', 'random'],
#                 'min_samples_split': [10,50,100]
#             }
#         },
#         'svc' : {
#             'model': SVC(),
#             'parameters': {
#                 'C': [10, 20, 50 ,100, 150],
#                 'kernel' : ['rbf','linear']
#             }
#         },
        'random forest' :{
            'model': RandomForestClassifier(),
            'parameters' :{
                'random_state': [5,10],
                'criterion' : [ 'gini', 'entropy'],
                'n_estimators': [1,5,10],

            }
        }
    }
    
    scores = []
    best_estimators = {}
    cv = ShuffleSplit(n_splits=10, test_size=0.25)
    
    for algos_name, config in algos.items():
        gs = GridSearchCV(config['model'], config['parameters'], cv=cv, return_train_score = False)
        gs.fit(X,Y)
        scores.append({
            'model' : algos_name,
            'best_score': gs.best_score_ ,
            'best_parameter': gs.best_params_
        })
        best_estimators[algos_name] = gs.best_estimator_
#     print(best_estimators)
    return pd.DataFrame(scores, columns= ['model', 'best_score', 'best_parameter'])

In [48]:
best_model_using_gridsearchCV(input_set, target_set)

model  best_score  \
0  linear regression    0.973913   
1      random forest    0.965217   

                                      best_parameter  
0  {'fit_intercept': True, 'random_state': 1, 'so...  
1  {'criterion': 'gini', 'n_estimators': 10, 'ran...

In [46]:
best_estimators

NameError: name 'best_estimators' is not defined

In [43]:
best_clf = best_estimators['random forest']

NameError: name 'best_estimators' is not defined

Saving the trained model

In [ ]:
!pip install joblib
import joblib 
# Save the model as a pickle in a file 
joblib.dump(best_clf, 'saved_model.pkl') 